xds/internal/balancer/cdsbalancer/cluster_handler.go

```diff
@@ -164,15 +182,19 @@ func createClusterNode(clusterName string, xdsClient xdsclient.XDSClient, topLev
 		topLevelHandler.parent.logger.Infof("CDS watch canceled on %v", clusterName)
 		cancel()
 	}
-	return c
+	topLevelHandler.createdClusters[clusterName] = c
 }
 
 // This function cancels the cluster watch on the cluster and all of it's
 // children.
 func (c *clusterNode) delete() {
-	c.cancelFunc()
-	for _, child := range c.children {
-		child.delete()
+	c.refCount--
+	if c.refCount == 0 {
+		c.cancelFunc()
+		delete(c.clusterHandler.createdClusters, c.clusterUpdate.ClusterName)
+		for _, child := range c.children {
+			c.clusterHandler.createdClusters[child].delete()
```

Comment:
I think that it wouldn't delete in 194, since if a node is a child of itself, the ref count wouldn't be zero. However, I added the nil check just in case. Added a test for node being a child of itself. I thought a lot about what it should do if it builds an update without erroring (due to ignoring dups) and is an empty cluster update. I decided to ignore, you can see the behavior in my TestNodeChildOfItself test.